sympy/polys/polytools.py

```diff
@@ -4404,13 +4404,33 @@ def degree(f, *gens, **args):
 
     """
     options.allowed_flags(args, ['gen', 'polys'])
+    f = sympify(f)
+    gen = args.pop('gen', 0)
+
+    if f.is_Poly:
+        if len(gens) == 0:
+            return sympify(f.degree(gen=gen))
+        f = f.as_expr()
 
     try:
-        F, opt = poly_from_expr(f, *gens, **args)
+        if f.is_Number:
+           F, opt = poly_from_expr(f, *gens, **args)
+        else:
+            F, opt = poly_from_expr(f, **args)
     except PolificationFailed as exc:
         raise ComputationFailed('degree', 1, exc)
 
-    return sympify(F.degree(opt.gen))
+    if len(gens) == 0:
+        gens = F.gens
+    if isinstance(gen, int):
+        try:
+            gen = gens[gen]
+        except IndexError:
+            raise PolynomialError("-%s <= gen < %s expected, got %s" %
+                                  (len(gens), len(gens), gen))
+    if gen in F.gens:
+        return sympify(F.degree(gen))
+    return S.Zero
```

Comment:
I'm not sure it's a good idea that degree and Poly.degree return different types. I think there are other functions that return ints. I see that you aren't changing anything, but perhaps this is an issue to look into.